src/physics/base/ValueGridBuilder.cc

```diff
@@ -0,0 +1,140 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file ValueGridBuilder.cc
+//---------------------------------------------------------------------------//
+#include "ValueGridBuilder.hh"
+
+#include <cmath>
+#include "base/SoftEqual.hh"
+#include "base/UniformGrid.hh"
+
+namespace celeritas
+{
+namespace
+{
+using SpanConstReal = ValueGridXsBuilder::SpanConstReal;
+//---------------------------------------------------------------------------//
+//// HELPER FUNCTIONS ////
+//---------------------------------------------------------------------------//
+bool is_contiguous_increasing(SpanConstReal first, SpanConstReal second)
+{
+    return first.size() >= 2 && second.size() >= 2 && first.front() > 0
+           && first.back() > first.front() && second.front() == first.back()
+           && second.back() > second.front();
+}
+
+bool has_same_log_spacing(SpanConstReal first, SpanConstReal second)
+{
+    auto calc_log_delta = [](SpanConstReal vec) {
+        return vec.back() / (vec.front() * (vec.size() - 1));
+    };
+    return soft_equal(calc_log_delta(first), calc_log_delta(second));
+}
+
+bool is_nonnegative(SpanConstReal vec)
+{
+    return std::all_of(
+        vec.begin(), vec.end(), [](real_type v) { return v >= 0; });
+}
+
+bool is_on_grid_point(real_type value, real_type lo, real_type hi, size_type size)
+{
+    if (value < lo || value > hi)
+        return false;
+
+    real_type delta = (hi - lo) / size;
+    return soft_zero(std::fmod(value - lo, delta));
+}
+
+//---------------------------------------------------------------------------//
+} // namespace
+
+//---------------------------------------------------------------------------//
+//! Default destructor
+ValueGridBuilder::~ValueGridBuilder() = default;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct XS arrays from imported data from Geant4.
+ */
+ValueGridXsBuilder
+ValueGridXsBuilder::from_geant(SpanConstReal lambda_energy,
+                               SpanConstReal lambda,
+                               SpanConstReal lambda_prim_energy,
+                               SpanConstReal lambda_prim)
+{
+    REQUIRE(is_contiguous_increasing(lambda_energy, lambda_prim_energy));
+    REQUIRE(has_same_log_spacing(lambda_prim, lambda_prim_energy));
+    REQUIRE(lambda.size() == lambda_energy.size());
+    REQUIRE(lambda_prim.size() == lambda_prim_energy.size());
+    REQUIRE(soft_equal(lambda.back(),
+                       lambda_prim.front() * lambda_prim_energy.front()));
+    REQUIRE(is_nonnegative(lambda) && is_nonnegative(lambda_prim));
+
+    // Concatenate the two XS vectors
+    std::vector<real_type> xs(lambda.size() + lambda_prim.size() - 1);
+    auto dst = std::copy(lambda.begin(), lambda.end(), xs.begin());
+    dst      = std::copy(lambda_prim.begin(), lambda_prim.end(), dst);
```

Comment:
Should this copy into `dst - 1`?

Reply:
Ooh, really good catch. The previous line should be `lambda.end() - 1`. This code is untested and can't reasonably be tested until I build out the value grid builder. Should I replace all this code with "UNREACHABLE" for now to indicate it's needed for the interface, until it can actually be tested?

Reply:
I figured, since the check would have caught it 😄. I think it's fine to leave it; if you want you could make a note that it's untested.